streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java

```diff
@@ -147,122 +187,124 @@ public File baseDir() {
+
+        // if the store name does not exist in the changelog map, it means the underlying store
+        // is not log enabled (including global stores), and hence it does not need to be restored
         final String topic = storeToChangelogTopic.get(storeName);
-        if (topic != null) {
-            final TopicPartition storePartition = new TopicPartition(topic, getPartition(topic));
+        if (topic != null && stateRestoreCallback != null) {
+            final TopicPartition storePartition = new TopicPartition(topic, taskId.partition);
+            storeMetadata.changelogPartition = storePartition;
+            storeMetadata.restoreCallback = stateRestoreCallback;
+            storeMetadata.recordConverter = converterForStore(store);
 
-            final RecordConverter recordConverter = converterForStore(store);
+            changelogReader.register(storePartition, this);
+        }
 
-            if (isStandby) {
-                log.trace("Preparing standby replica of persistent state store {} with changelog topic {}", storeName, topic);
+        stores.put(storeName, storeMetadata);
 
-                restoreCallbacks.put(topic, stateRestoreCallback);
-                recordConverters.put(topic, recordConverter);
-            } else {
-                final Long restoreCheckpoint = store.persistent() ? initialLoadedCheckpoints.get(storePartition) : null;
-                if (restoreCheckpoint != null) {
-                    checkpointFileCache.put(storePartition, restoreCheckpoint);
-                }
-                log.trace("Restoring state store {} from changelog topic {} at checkpoint {}", storeName, topic, restoreCheckpoint);
-
-                final StateRestorer restorer = new StateRestorer(
-                    storePartition,
-                    new CompositeRestoreListener(stateRestoreCallback),
-                    restoreCheckpoint,
-                    offsetLimit(storePartition),
-                    store.persistent(),
-                    storeName,
-                    recordConverter
-                );
-
-                changelogReader.register(restorer);
-            }
-            changelogPartitions.add(storePartition);
+        log.debug("Registered state store {} to its state manager", storeName);
+    }
+
+    @Override
+    public StateStore getStore(final String name) {
+        if (stores.containsKey(name)) {
+            return stores.get(name).stateStore;
+        } else {
+            return null;
         }
+    }
 
-        registeredStores.put(storeName, Optional.of(store));
+    Collection<TopicPartition> changelogPartitions() {
+        return changelogOffsets().keySet();
     }
 
     @Override
-    public void reinitializeStateStoresForPartitions(final Collection<TopicPartition> partitions,
-                                                     final InternalProcessorContext processorContext) {
-        StateManagerUtil.reinitializeStateStoresForPartitions(log,
```

Comment:
Haha you got me! I left it intentionally to guide me how to implement the exception handling logic on the thread level just so that I do not need to check out trunk and compare :P If you insist I can remove it in this PR too.

Reply:
No I'm not insisting, help yourself :)